lib/Sema/CSDiag.cpp

```diff
@@ -4361,8 +4361,13 @@ bool TypeChecker::diagnoseSelfAssignment(const Expr *E) {
   if (!AE)
     return false;
 
+  auto src = AE->getSrc();
+  auto loadSrc = dyn_cast<LoadExpr>(src);
```

Comment:
If the right hand side is not an lvalue (e.g., you are assigning the result of a function call) this returns null and you will hit a null pointer deref on the next line. You should unwrap the LoadExpr if one is there, but otherwise keep going.